### ISSUE TITLE
🐛 [firebase_messaging] _firebaseMessagingBackgroundHandler is not being executed on background notification received

### ISSUE THREAD

**Issue by krupikivan** [firebase/flutterfire#9932]
## Bug report

**Describe the bug**
_firebaseMessagingBackgroundHandler is not being excited on background notification received
```
D/FLTFireMsgReceiver(23619): broadcast received for message. 
E/FLTFireBGExecutor(23619): Background isolate already started. 
I/FLTFireMsgService(23619): Service has not yet started, messages will be queued.
```

### Expected behavior

Should execute _firebaseMessagingBackgroundHandler function on background notification

### Sample project

My Code :

```
@pragma('vm:entry-point')
Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);
  print(message.toString());
}

Future main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);
  FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);
  await setupGlobalLocator();
  runZonedGuarded(() {
    runApp(
      MyApp(),
    );
  }, (error, stackTrace) {
    FirebaseCrashlytics.instance.recordError(error, stackTrace);
  });
}
```

---

## Additional context

The Notification is shown in the app

---
### Flutter Doctor

```
[✓] Flutter (Channel stable, 3.3.4, on macOS 13.0 22A380 darwin-arm, locale en-US)
    • Flutter version 3.3.4 on channel stable at /Users/user/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eb6d86ee27 (6 weeks ago), 2022-10-04 22:31:45 -0700
    • Engine revision c08d7d5efc
    • Dart version 2.18.2
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/user/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] IntelliJ IDEA Community Edition (version 2021.3.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] VS Code (version 1.66.2)
    • VS Code at /Users/user/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (4 available)
    • Pixel 2 (mobile) • HT83G1A00269       • android-arm64  • Android 11 (API 30)
    • SM T540 (mobile) • 192.168.1.152:5555 • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos              • darwin-arm64   • macOS 13.0 22A380 darwin-arm
    • Chrome (web)     • chrome             • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
  firebase_messaging: ^14.1.0
  firebase_crashlytics: ^3.0.5
  firebase_core: ^2.2.0
  flutter_local_notifications: ^12.0.3
  permission_handler: ^10.2.0
```

</details>

---

**Comment by Kangyeongwoo** [comment]
I meet same problem only Android Background , ios working well

**Comment by samueldelamotte** [comment]
I am seeing the exact same issue on Android background messages...
firebase_messaging: ^14.1.0

`D/FLTFireMsgReceiver(12436): broadcast received for message
E/FLTFireBGExecutor(12436): Background isolate already started.
I/FLTFireMsgService(12436): Service has not yet started, messages will be queued.`

**Comment by pleky** [comment]
I am facing the same issue...
firebase_messaging: ^14.1.0

**Comment by faviansyahap** [comment]
Hello, I'm facing same issue too on Android.

**Comment by kelpo123** [comment]
Ios working well, For android I didn't find any solution for this problem.

**Comment by darshankawar** [comment]
@krupikivan and everyone facing this issue: Can anybody confirm if this behavior is occuring using previous version of plugin too or not ? https://pub.dev/packages/firebase_messaging/changelog#1404

If not, and occurs only using latest plugin version, ie, `14.1.0`, please provide complete verbose log for further debugging.

**Comment by Abhilashkv013** [comment]
D/FLTFireMsgReceiver(27031): broadcast received for message
E/FLTFireBGExecutor(27031): Background isolate already started.
I/FLTFireMsgService(27031): Service has not yet started, messages will be queued.

**Comment by Abhilashkv013** [comment]
The main function called twice when use onBackgroundMessage

FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);

**Comment by krunaladit** [comment]
I am facing same issue in Android.

**Comment by SvenSchoene** [comment]
> @krupikivan and everyone facing this issue: Can anybody confirm if this behavior is occuring using previous version of plugin too or not ? https://pub.dev/packages/firebase_messaging/changelog#1404
> 
> If not, and occurs only using latest plugin version, ie, `14.1.0`, please provide complete verbose log for further debugging.

Unfortunately we're having the same issue with 14.04..

**Comment by Abhilashkv013** [comment]
we're having the same issue with previous versions.

**Comment by anton-brass** [comment]
same here on Android Emulator.

Is there a 14.x version working?

**Comment by SvenSchoene** [comment]
We reverted back to:

```
firebase_core: ^1.20.1
firebase_messaging: ^12.0.2
```

(The `firebase_messaging: 13*` versions didn't work for us for other reasons, see: https://github.com/firebase/flutterfire/issues/9806 .)

**Comment by SvenSchoene** [comment]
@krupikivan  I'm testing it now. Notifications worked up until iOS 15 so far.

But since you are asking I'm suspecting that iOS 16 might be an issue? My iPhone is upgrading as I'm writing this. I'll report back.

---

> Error: Member not found: 'FirebaseAppPlatform.verifyExtends'

I have that issue if I upgrade `firebase_core` to a 2.* version, see: https://github.com/firebase/flutterfire/issues/9806
Try downgrading `firebase_core` as well. That should work.

**Comment by Lyokone** [comment]
Hello, thanks for the report, I'm unable to reproduce it with the latest version of firebase messaging and the sample code.
```
2022-11-15 14:59:12.846  8897-8897  FLTFireMsgReceiver      io....ns.firebase.messaging.example  D  broadcast received for message
2022-11-15 14:59:12.861  8897-8897  FLTFireBGExecutor       io....ns.firebase.messaging.example  E  Background isolate already started.
2022-11-15 14:59:12.863  8897-8921  FLTFireMsgService       io....ns.firebase.messaging.example  I  Service has not yet started, messages will be queued.
```
I have the same messages, but the notification is showing as expected. 

The reported error is about Android, please keep the thread clean, and open another issue for iOS if needed.

**Comment by Lyokone** [comment]
Since [version 13 you need new permissions](https://pub.dev/packages/firebase_messaging/changelog#1300), did you follow the instructions:
```
BREAKING FEAT: firebase_messaging now includes this permission: Manifest.permission.POST_NOTIFICATIONS in its AndroidManifest.xml file which requires updating your android/app/build.gradle to target API level 33.
```

**Comment by krupikivan** [comment]
@Lyokone Could you please provide your sample code? I'm still having same Issue on Android

**Comment by Lyokone** [comment]
You can find the complete sample here: https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/example/lib/main.dart

**Comment by krupikivan** [comment]
Same sample code, get the firebase token, go to messaging cloud send notification and notification is not showing in Android with same console log

**Comment by Lyokone** [comment]
@krupikivan, as said before, it works in all 3 states for me (foreground, background, and killed).
Did you checked this message: https://github.com/firebase/flutterfire/issues/9932#issuecomment-1315356027

Please send a complete example in the form of a Github repo so I can check this (with your Manifest)

**Comment by krupikivan** [comment]
@Lyokone 
Inside manifest: <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
build.gradle: compileSdkVersion 33

**Comment by SvenSchoene** [comment]
Ok, we can work this out without me having to do a full reproduction repo. We'll see.

I did _not_ have the permission previously. I added that now.
Result: It still does not work.

Let me give you a bit of additional info and screenshots:

`android\app\src\main\AndroidManifest.xml` looks like this:
```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="***">
    <application android:label="***"
        android:name="${applicationName}"
        android:icon="@mipmap/ic_launcher">
        <activity android:name=".MainActivity"
            android:exported="true"
            android:launchMode="singleTop"
            android:theme="@style/LaunchTheme"
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
            android:hardwareAccelerated="true"
            android:windowSoftInputMode="adjustResize">
            <!-- Specifies an Android theme to apply to this Activity as soon as
                 the Android process has started. This theme is visible to the user
                 while the Flutter UI initializes. After that, this theme continues
                 to determine the Window background behind the Flutter UI. -->
            <meta-data android:name="io.flutter.embedding.android.NormalTheme"
                android:resource="@style/NormalTheme" />
            <intent-filter>
                <action android:name="android.intent.action.MAIN"/>
                <category android:name="android.intent.category.LAUNCHER"/>
            </intent-filter>
        </activity>
        <!-- Don't delete the meta-data below.
             This is used by the Flutter tool to generate GeneratedPluginRegistrant.java -->
        <meta-data android:name="flutterEmbedding"
            android:value="2" />
    </application>

    <uses-permission android:name="android.permission.INTERNET"/>
    <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
</manifest>
```

![image](https://user-images.githubusercontent.com/7631673/201948595-20f4d899-81ea-4e51-8449-616f491228cc.png)

My dependencies look like this:
```yml
  firebase_core: ^2.2.0
  firebase_messaging: ^14.1.0
```

![image](https://user-images.githubusercontent.com/7631673/201948775-20c631e6-0f2a-4c0a-a969-f763c9a066c5.png)

The Android version of my emulator is 12.

![android_emulator_version_1](https://user-images.githubusercontent.com/7631673/201949123-df0dcd10-cc36-4567-9132-2ec6c2e4603b.gif)

When I receive a notification in the background I get the following:
```
D/FLTFireMsgReceiver( 9494): broadcast received for message
E/FLTFireBGExecutor( 9494): Background isolate already started.
I/FLTFireMsgService( 9494): Service has not yet started, messages will be queued.
```

![image](https://user-images.githubusercontent.com/7631673/201949638-b4e9095d-7172-4b74-b89f-08642addd04c.png)

I hope we figure this out.

**Comment by SvenSchoene** [comment]
When I send a push notification from our backend, `adb logcat` shows the following:

```
11-15 14:54:40.834  1121  6992 W ChimeraUtils: Module com.google.android.gms.gcm has empty metadata display_name_string_id
11-15 14:54:40.837  9494  9494 D FLTFireMsgReceiver: broadcast received for message
11-15 14:54:40.843  9494  9494 E FLTFireBGExecutor: Background isolate already started.
11-15 14:54:40.845  9494  9516 I FLTFireMsgService: Service has not yet started, messages will be queued.
11-15 14:54:40.851  9494  9494 D FLTFireMsgReceiver: broadcast received for message
11-15 14:54:40.856  9494  9494 E FLTFireBGExecutor: Background isolate already started.
11-15 14:54:40.858  9494  9516 I FLTFireMsgService: Service has not yet started, messages will be queued.
```

![image](https://user-images.githubusercontent.com/7631673/201950824-360ec275-547c-4d2a-83bc-06e25fd1fe6f.png)

**Comment by Lyokone** [comment]
As stated [here](https://github.com/firebase/flutterfire/issues/9932#issuecomment-1315351425), those messages are not indicative of the notification not being received.

Can you try the [messaging sample](https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/example/lib/main.dart) app to see if it's related to your device? You have information on how to run the example app in the `main.dart`

**Comment by krupikivan** [comment]
@Lyokone I just created new flutter project and Background notification works in Android looks like we forgot to update some of the dependencies
Here is the sample code
https://github.com/krupikivan/firebase_notification_background

**Comment by Lyokone** [comment]
@krupikivan Glad to hear that, thanks for updating :)

**Comment by SvenSchoene** [comment]
Regarding setting up the sample app: I'm struggling with step 4:

> 4. Run `flutterfire configure` in the example/ directory to setup your app with your Firebase project.

Where is the `flutterfire`-binary supposed come from?

---

In any case: If this works for krupikivan with his sample application (and it works for you, Lyokone, as well), then I guess we can assume that it works in general. The question then is: What do I (and the other people in this thread) need to do to migrate our existing projects to the new version so that it works?

I'll report back if I find the solution.

**Comment by krupikivan** [comment]
@SvenSchoene  Double check your dependencies like 
`implementation platform('com.google.firebase:firebase-bom:31.0.3')`

**Comment by krupikivan** [comment]
@Lyokone IM able to receive the notification but I'm not able to run anything inside `_firebaseMessagingBackgroundHandler`
In mi case I want to dispatch a `FlutterLocalNotification` inside `_firebaseMessagingBackgroundHandler` function and is not working. However `FirebaseMessaging.onMessageOpenedApp` works if I tap the notification

**Comment by SvenSchoene** [comment]
@krupikivan Thank you for the suggestion!

Unfortunately, that did not work. Even after a `flutter clean`.
Also: I can't find it anywhere in the docs that that dependency has to be added manually. How did you get the idea to add that?

---
Regarding your next reply: How do you get a notification if the background handler (and therefore FlutterLocalNotification) is not executed? Or is the notification that you can tap from Firebase itself?

**Comment by krupikivan** [comment]
@SvenSchoene  In my case notification is showing in my phone, but like you `_firebaseMessagingBackgroundHandler` is not executed.
When I created a new Firebase project through the Firebase console right after getting my google-service.json one of the this was one of the dependencies I had to migrate

**Comment by krupikivan** [comment]
@Lyokone Are you able to execute the peace of code inside `_firebaseMessagingBackgroundHandler`?

**Comment by apaolino** [comment]
Experiencing the same, and I'm having a huge drop in received data messages in previous firebase_messaging versions also
https://github.com/firebase/flutterfire/issues/9930#event-7811522455

**Comment by krupikivan** [comment]
@Lyokone  I had to reopen the issue BTW I changed the title to be specific

**Comment by sarbagyastha** [comment]
I confirmed that the issue only exist for `firebase_messaging: 14.1.0`, all version prior to it works.
So, most probably https://github.com/firebase/flutterfire/pull/9867 caused it.

**Comment by darshankawar** [comment]
Using the [plugin example](https://github.com/firebase/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example) and running on Android, although I am getting same logs as reported, but notification is received:

```
broadcast received for message
Background isolate already started.
Service has not yet started, messages will be queued.
```

But looking at the comments received, I am going ahead and labeling this for further insights and tracking for team to look into.

May be related : https://github.com/firebase/flutterfire/issues/9912

**Comment by SvenSchoene** [comment]
> 

I guess at this point the question is not whether the notification is received, but whether the method `_firebaseMessagingBackgroundHandler()` is executed: https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/example/lib/main.dart#L47-L54

It is set up here in the example: https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/example/lib/main.dart#L123

I couldn't get the example to run. But you did. So could you run it again and check whether you see the output for this call from line 53?

```dart
print('Handling a background message ${message.messageId}');
```

**Comment by webranch-pl** [comment]
Is something already set? Maybe could you provie last working version, which is tested and recommended? I mean firebase_core and firebase_messaging plugins?

**Comment by SvenSchoene** [comment]
Could it be that there is actually nothing wrong with version `14.1.0`, but much rather the issue is what I stumbled upon here: https://github.com/firebase/flutterfire/issues/9951#issuecomment-1318548853

TLDR: Adding `@pragma('vm:entry-point')` to the background message handler fixes the issue?

I do not have time to test `14.1.0` with this at the moment. I'm scrambling to get everything ready for my 10 day vacation. 😅 

But I had to share this.

**Comment by Abhilashkv013** [comment]
I still have the same problem

**Comment by SvenSchoene** [comment]
@Abhilashkv013 Try changing your version string from `^14.0.4` to `14.0.4`. The `^` might still resolve to `14.1.0` (I must admit that I always forget semantic versioning details... 😬 ).

ALSO: Remove your `pubspec.lock` and run `flutter pub get` again. I was finding myself in a situation where `14.1.0` was still the resolved version, even after changing the version to `^14.0.4` or even `14.0.4`.

(Maybe `^14.0.4` + removing the `pubspec.lock` file is all it takes -- meaning the `^` can stay.)

Otherwise your code looks fine to me.

If you are testing on Android 13 make sure to add `<uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>` to your `AndroidManifest.xml` as well.

**Comment by Abhilashkv013** [comment]
> @Abhilashkv013 Try changing your version string from `^14.0.4` to `14.0.4`. The `^` might still resolve to `14.1.0` (I must admit that I always forget semantic versioning details... 😬 ).
> 
> ALSO: Remove your `pubspec.lock` and run `flutter pub get` again. I was finding myself in a situation where `14.1.0` was still the resolved version, even after changing the version to `^14.0.4` or even `14.0.4`.
> 
> (Maybe `^14.0.4` + removing the `pubspec.lock` file is all it takes -- meaning the `^` can stay.)
> 
> Otherwise your code looks fine to me.
> 
> If you are testing on Android 13 make sure to add `<uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>` to your `AndroidManifest.xml` as well.

Thanks Working

**Comment by prakasharyaman** [comment]
Just upgraded to and it's working.
`  firebase_messaging: ^14.0.4`
